File: impl/php/src/BasicSuffixProfile.php
<?php

declare(strict_types=1);

namespace Makrell\Formats;

final class BasicSuffixProfile
{
    public static function applyString(string $value, string $suffix): mixed
    {
        if ($suffix === '') {
            return $value;
        }

        return match ($suffix) {
            'dt' => ['value' => $value, 'suffix' => $suffix],
            'bin' => self::parseBase($value, 2, 'binary'),
            'oct' => self::parseBase($value, 8, 'octal'),
            'hex' => self::parseBase($value, 16, 'hexadecimal'),
            default => throw new MakrellFormatException("Unsupported basic suffix profile string suffix '{$suffix}'."),
        };
    }

    public static function applyNumber(string $value, string $suffix): int|float
    {
        if ($suffix === '') {
            if (str_contains($value, '.') || str_contains($value, 'e') || str_contains($value, 'E')) {
                return (float) $value;
            }
            return (int) $value;
        }

        if (str_contains($value, '.') || str_contains($value, 'e') || str_contains($value, 'E')) {
            return self::applyFloatSuffix((float) $value, $suffix);
        }

        $base = (int) $value;
        return match ($suffix) {
            'k' => $base * 1_000,
            'M' => $base * 1_000_000,
            'G' => $base * 1_000_000_000,
            'T' => $base * 1_000_000_000_000,
            'P' => $base * 1_000_000_000_000_000,
            'E' => $base * 1_000_000_000_000_000_000,
            default => self::applyFloatSuffix((float) $base, $suffix),
        };
    }

    public static function splitNumericLiteralSuffix(string $text): ?array
    {
        $length = strlen($text);
        for ($boundary = $length; $boundary > 0; $boundary--) {
            $value = substr($text, 0, $boundary);
            $suffix = substr($text, $boundary);
            if ($suffix !== '' && !self::isSuffixIdentifier($suffix)) {
                continue;
            }
            if (filter_var($value, FILTER_VALIDATE_INT) !== false || is_numeric($value)) {
                return [$value, $suffix];
            }
        }
        return null;
    }

    public static function isTaggedString(mixed $value): bool
    {
        return is_array($value)
            && array_keys($value) === ['value', 'suffix']
            && is_string($value['value'])
            && is_string($value['suffix']);
    }

    private static function parseBase(string $value, int $base, string $label): int
    {
        $parsed = intval($value, $base);
        if (strtolower(base_convert((string) $parsed, 10, $base)) !== strtolower(ltrim($value, '0')) && $value !== '0') {
            throw new MakrellFormatException("Invalid {$label} literal '{$value}'.");
        }
        return $parsed;
    }

    private static function applyFloatSuffix(float $base, string $suffix): float
    {
        return match ($suffix) {
            'k' => $base * 1e3,
            'M' => $base * 1e6,
            'G' => $base * 1e9,
            'T' => $base * 1e12,
            'P' => $base * 1e15,
            'E' => $base * 1e18,
            'pi' => $base * M_PI,
            'tau' => $base * (2 * M_PI),
            'deg' => $base * (M_PI / 180.0),
            'e' => $base * exp(1),
            default => throw new MakrellFormatException("Unsupported basic suffix profile numeric suffix '{$suffix}'."),
        };
    }

    private static function isSuffixIdentifier(string $suffix): bool
    {
        return preg_match('/^[A-Za-z_][A-Za-z0-9_]*$/', $suffix) === 1;
    }
}

File: impl/rust/src/basic_suffix_profile.rs
use crate::MakrellFormatError;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum BasicSuffixValue {
    String(String),
    Int(i64),
    Float(f64),
    TaggedString { value: String, suffix: String },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BasicSuffixLiteralKind {
    String,
    Number,
}

pub fn split_numeric_literal_suffix(text: &str) -> Option<(&str, &str)> {
    let boundaries: Vec<usize> = text
        .char_indices()
        .map(|(index, _)| index)
        .chain(std::iter::once(text.len()))
        .collect();

    for &boundary in boundaries.iter().rev() {
        let (value, suffix) = text.split_at(boundary);
        if value.is_empty() {
            continue;
        }
        if !suffix.is_empty() && !is_suffix_identifier(suffix) {
            continue;
        }
        if value.parse::<i64>().is_ok() || value.parse::<f64>().is_ok() {
            return Some((value, suffix));
        }
    }
    None
}

pub fn apply_basic_suffix_profile(
    kind: BasicSuffixLiteralKind,
    value: &str,
    suffix: &str,
) -> Result<BasicSuffixValue, MakrellFormatError> {
    match kind {
        BasicSuffixLiteralKind::String => {
            if suffix.is_empty() {
                return Ok(BasicSuffixValue::String(value.to_string()));
            }
            match suffix {
                "dt" => Ok(BasicSuffixValue::TaggedString {
                    value: value.to_string(),
                    suffix: suffix.to_string(),
                }),
                "bin" => Ok(BasicSuffixValue::Int(
                    i64::from_str_radix(value, 2)
                        .map_err(|_| MakrellFormatError::new(format!("Invalid binary literal '{value}'.")))?,
                )),
                "oct" => Ok(BasicSuffixValue::Int(
                    i64::from_str_radix(value, 8)
                        .map_err(|_| MakrellFormatError::new(format!("Invalid octal literal '{value}'.")))?,
                )),
                "hex" => Ok(BasicSuffixValue::Int(
                    i64::from_str_radix(value, 16)
                        .map_err(|_| MakrellFormatError::new(format!("Invalid hexadecimal literal '{value}'.")))?,
                )),
                _ => Err(MakrellFormatError::new(format!(
                    "Unsupported basic suffix profile string suffix '{suffix}'."
                ))),
            }
        }
        BasicSuffixLiteralKind::Number => apply_basic_number_suffix(value, suffix),
    }
}

fn apply_basic_number_suffix(value: &str, suffix: &str) -> Result<BasicSuffixValue, MakrellFormatError> {
    if suffix.is_empty() {
        if value.contains('.') || value.contains('e') || value.contains('E') {
            return Ok(BasicSuffixValue::Float(
                value
                    .parse::<f64>()
                    .map_err(|_| MakrellFormatError::new(format!("Invalid numeric literal '{value}'.")))?,
            ));
        }
        return Ok(BasicSuffixValue::Int(
            value
                .parse::<i64>()
                .map_err(|_| MakrellFormatError::new(format!("Invalid numeric literal '{value}'.")))?,
        ));
    }

    if value.contains('.') || value.contains('e') || value.contains('E') {
        let base = value
            .parse::<f64>()
            .map_err(|_| MakrellFormatError::new(format!("Invalid numeric literal '{value}'.")))?;
        return apply_basic_float_suffix(base, suffix);
    }

    let base = value
        .parse::<i64>()
        .map_err(|_| MakrellFormatError::new(format!("Invalid numeric literal '{value}'.")))?;
    match suffix {
        "k" => Ok(BasicSuffixValue::Int(base * 1_000)),
        "M" => Ok(BasicSuffixValue::Int(base * 1_000_000)),
        "G" => Ok(BasicSuffixValue::Int(base * 1_000_000_000)),
        "T" => Ok(BasicSuffixValue::Int(base * 1_000_000_000_000)),
        "P" => Ok(BasicSuffixValue::Int(base * 1_000_000_000_000_000)),
        "E" => Ok(BasicSuffixValue::Int(base * 1_000_000_000_000_000_000)),
        _ => apply_basic_float_suffix(base as f64, suffix),
    }
}

fn is_suffix_identifier(text: &str) -> bool {
    let mut chars = text.chars();
    match chars.next() {
        Some(first) if first.is_ascii_alphabetic() || first == '_' => {}
        _ => return false,
    }
    chars.all(|ch| ch.is_ascii_alphanumeric() || ch == '_')
}

fn apply_basic_float_suffix(base: f64, suffix: &str) -> Result<BasicSuffixValue, MakrellFormatError> {
    let value = match suffix {
        "k" => base * 1e3,
        "M" => base * 1e6,
        "G" => base * 1e9,
        "T" => base * 1e12,
        "P" => base * 1e15,
        "E" => base * 1e18,
        "pi" => base * std::f64::consts::PI,
        "tau" => base * std::f64::consts::TAU,
        "deg" => base * (std::f64::consts::PI / 180.0),
        "e" => base * std::f64::consts::E,
        _ => {
            return Err(MakrellFormatError::new(format!(
                "Unsupported basic suffix profile numeric suffix '{suffix}'."
            )))
        }
    };
    Ok(BasicSuffixValue::Float(value))
}
